File: scripts/match-reviews-to-cities.js
#!/usr/bin/env node
/**
 * Match Google Reviews to Cities via Jobber
 * Searches Jobber for reviewer names and extracts their city
 */

const fs = require('fs');
const path = require('path');

// Load credentials
const creds = JSON.parse(fs.readFileSync('/Users/jarvis/clawd/jobber_credentials.json', 'utf8'));

// Load reviews
const reviews = JSON.parse(fs.readFileSync('/Users/jarvis/clawd/scws-gmb/reviews.json', 'utf8'));

async function searchJobberClient(name) {
  const query = `
    query SearchClients($searchTerm: String!) {
      clients(searchTerm: $searchTerm, first: 5) {
        nodes {
          id
          name
          firstName
          lastName
          billingAddress {
            city
            street
            postalCode
          }
        }
      }
    }
  `;

  try {
    const response = await fetch(creds.api_endpoint, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${creds.access_token}`,
        'X-JOBBER-GRAPHQL-VERSION': creds.graphql_version
      },
      body: JSON.stringify({
        query,
        variables: { searchTerm: name }
      })
    });

    const data = await response.json();
    if (data.errors) {
      console.error(`Error searching for ${name}:`, data.errors[0].message);
      return null;
    }
    return data.data?.clients?.nodes || [];
  } catch (err) {
    console.error(`Failed to search for ${name}:`, err.message);
    return null;
  }
}

function extractCity(client) {
  if (client.billingAddress?.city) {
    return client.billingAddress.city;
  }
  return null;
}

function cleanName(displayName) {
  // Remove nicknames in parentheses, emojis, etc.
  return displayName
    .replace(/\(.*?\)/g, '')
    .replace(/[^\w\s'-]/g, '')
    .trim();
}

function nameMatch(reviewerName, clientName) {
  const rn = reviewerName.toLowerCase().trim();
  const cn = clientName.toLowerCase().trim();
  
  // Exact match
  if (rn === cn) return true;
  
  // Check if reviewer name parts are in client name
  const rParts = rn.split(/\s+/);
  const cParts = cn.split(/\s+/);
  
  // At least first and last name match
  if (rParts.length >= 2 && cParts.length >= 2) {
    const firstMatch = rParts[0] === cParts[0];
    const lastMatch = rParts[rParts.length - 1] === cParts[cParts.length - 1];
    if (firstMatch && lastMatch) return true;
  }
  
  // First name match for single names
  if (rParts.length === 1 && rParts[0] === cParts[0]) return true;
  
  return false;
}

async function main() {
  console.log('Matching Google Reviews to Cities via Jobber\n');
  
  // Collect all 5-star reviews with comments
  const fiveStarReviews = [];
  for (const location of reviews) {
    for (const review of location.reviews) {
      if (review.starRating === 'FIVE' && review.comment) {
        fiveStarReviews.push({
          name: review.reviewer.displayName,
          text: review.comment,
          date: review.createTime.split('T')[0],
          location: location.address.locality
        });
      }
    }
  }
  
  console.log(`Found ${fiveStarReviews.length} 5-star reviews with comments\n`);
  
  const reviewsByCity = {};
  const matched = [];
  const unmatched = [];
  
  // Process each review
  for (const review of fiveStarReviews) {
    const cleanedName = cleanName(review.name);
    console.log(`Searching for: ${cleanedName}`);
    
    const clients = await searchJobberClient(cleanedName);
    
    if (clients && clients.length > 0) {
      // Find best match
      let bestMatch = null;
      for (const client of clients) {
        if (nameMatch(cleanedName, client.name)) {
          bestMatch = client;
          break;
        }
      }
      
      if (bestMatch) {
        const city = extractCity(bestMatch);
        if (city) {
          const normalizedCity = city.trim();
          console.log(`  ✓ Matched to ${bestMatch.name} in ${normalizedCity}`);
          
          if (!reviewsByCity[normalizedCity]) {
            reviewsByCity[normalizedCity] = [];
          }
          
          reviewsByCity[normalizedCity].push({
            name: review.name,
            text: review.text,
            date: review.date
          });
          
          matched.push({ reviewer: review.name, client: bestMatch.name, city: normalizedCity });
        } else {
          console.log(`  ~ Matched ${bestMatch.name} but no city found`);
          unmatched.push(review.name);
        }
      } else {
        console.log(`  ✗ No name match in results`);
        unmatched.push(review.name);
      }
    } else {
      console.log(`  ✗ No results`);
      unmatched.push(review.name);
    }
    
    // Rate limiting
    await new Promise(r => setTimeout(r, 200));
  }
  
  // Save results
  const outputPath = '/Users/jarvis/clawd/scws-website/scripts/reviews-by-city.json';
  fs.writeFileSync(outputPath, JSON.stringify(reviewsByCity, null, 2));
  
  console.log('\n' + '='.repeat(50));
  console.log(`\n✅ RESULTS:`);
  console.log(`   Matched: ${matched.length} reviews`);
  console.log(`   Unmatched: ${unmatched.length} reviews`);
  console.log(`\n📍 Cities with testimonials:`);
  
  const sortedCities = Object.entries(reviewsByCity)
    .sort((a, b) => b[1].length - a[1].length);
  
  for (const [city, cityReviews] of sortedCities) {
    console.log(`   ${city}: ${cityReviews.length} reviews`);
  }
  
  console.log(`\n💾 Saved to: ${outputPath}`);
}

main().catch(console.error);

File: scripts/match-reviews-to-cities.py
#!/usr/bin/env python3
"""Match Google reviews to Jobber clients to get cities."""
import json, requests, time, sys

with open('/Users/jarvis/clawd/jobber_credentials.json') as f:
    creds = json.load(f)

TOKEN = creds['access_token']
API = 'https://api.getjobber.com/api/graphql'
HEADERS = {
    'Authorization': f'Bearer {TOKEN}',
    'Content-Type': 'application/json',
    'X-JOBBER-GRAPHQL-VERSION': '2025-01-20'
}

def search_client(name):
    query = '{clients(searchTerm: "%s", first: 5) { nodes { firstName lastName billingAddress { city } properties { address { city } } } } }' % name.replace('"', '\\"')
    r = requests.post(API, json={'query': query}, headers=HEADERS, timeout=15)
    if r.status_code == 200:
        data = r.json().get('data', {})
        return data.get('clients', {}).get('nodes', []) if data else []
    return []

# Load reviews
with open('/Users/jarvis/clawd/scws-gmb/reviews.json') as f:
    data = json.load(f)

five_star = []
for loc in data:
    for r in loc.get('reviews', []):
        if r.get('starRating') == 'FIVE' and r.get('comment'):
            five_star.append({
                'name': r.get('reviewer', {}).get('displayName', ''),
                'text': r.get('comment', ''),
                'date': r.get('createTime', '')[:10],
            })

print(f"Processing {len(five_star)} 5-star reviews with comments...", flush=True)

SKIP = {'Phoenix', 'Bob Nunya', 'Ricardo', 'Maggie', 'Nancy', 'fromelkriver',
        'Sand to Sea Videos', 'Capital & Influence Inc', 'Birdsong Retreat',
        "Ledezma #6", "George's Perez", "Nadia S", "Maria M", "Sabrina N", 
        "Jarrod P", "sister haero"}

matched = {}
unmatched = []

for i, review in enumerate(five_star):
    name = review['name']
    if name in SKIP:
        print(f"  [{i+1}] SKIP {name}", flush=True)
        unmatched.append(review)
        continue
    
    search_name = name.split('(')[0].strip()
    search_name = search_name.replace("Keep n' up with ", "")
    
    clients = search_client(search_name)
    time.sleep(0.25)
    
    city = None
    if clients:
        c = clients[0]
        props = c.get('properties', [])
        if props and props[0].get('address', {}).get('city'):
            city = props[0]['address']['city']
        elif c.get('billingAddress', {}).get('city'):
            city = c['billingAddress']['city']
    
    if city:
        city = city.strip().title()
        matched.setdefault(city, []).append({
            'name': review['name'],
            'text': review['text'],
            'date': review['date']
        })
        print(f"  [{i+1}] ✅ {name} → {city}", flush=True)
    else:
        unmatched.append(review)
        print(f"  [{i+1}] ❌ {name}", flush=True)

output_path = '/Users/jarvis/clawd/scws-website/scripts/reviews-by-city.json'
with open(output_path, 'w') as f:
    json.dump(matched, f, indent=2)

print(f"\n{'='*50}", flush=True)
print(f"Matched: {sum(len(v) for v in matched.values())} reviews", flush=True)
print(f"Unmatched: {len(unmatched)} reviews", flush=True)
print(f"Cities: {len(matched)}", flush=True)
print(f"\nCities with testimonials:", flush=True)
for city, reviews in sorted(matched.items(), key=lambda x: -len(x[1])):
    print(f"  {city}: {len(reviews)} reviews", flush=True)
